File: src/shapes/Rectangle.java
package shapes;

public class Rectangle extends java.lang.Object {
    private double length = 1.0f;
    private double width = 1.0f;

    public Rectangle() {
    }

    public Rectangle(double size) {
        this.length = size;
        this.width = size;
    }

    public Rectangle(double length, double width) {
        this.length = length;
        this.width = width;
    }

    public double getArea() {
        return this.getLength() * this.getWidth();
    }

    public double getPerimeter() {
        return (2 * this.getLength()) + (2 * this.getWidth());
    }

    public double getLength() {
        return this.length;
    }

    public double getWidth() {
        return this.width;
    }

    public void setLength(double length) {
        this.length = length;
    }

    public void setWidth(double width) {
        this.width = width;
    }

    public String toString() {
        return "rectangle with length " + this.getLength() + ", width " + this.getWidth();
    }
}

File: rust/src/shapes/rectangle.rs
use std::fmt::Display;

pub struct Rectangle(f64, f64);

impl Rectangle {
    pub fn new(length: f64, width: f64) -> Self {
        Self(length, width)
    }

    pub fn square(size: f64) -> Self {
        Self(size, size)
    }

    pub fn get_area(&self) -> f64 {
        self.get_length() * self.get_width()
    }

    pub fn get_perimeter(&self) -> f64 {
        (2.0 * self.get_length()) + (2.0 * self.get_width())
    }

    pub fn get_length(&self) -> f64 {
        self.0
    }

    pub fn get_width(&self) -> f64 {
        self.1
    }

    pub fn set_length(&mut self, length: f64) {
        self.0 = length
    }

    pub fn set_width(&mut self, width: f64) {
        self.1 = width
    }
}

impl Default for Rectangle {
    fn default() -> Self {
        Self(1.0, 1.0)
    }
}

impl Display for Rectangle {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "rectangle with length {}, width {}",
            self.get_length(),
            self.get_width()
        )
    }
}
